source/common/upstream/subset_lb.cc

```diff
@@ -141,6 +211,65 @@ HostConstSharedPtr SubsetLoadBalancer::chooseHost(LoadBalancerContext* context)
   return nullptr;
 }
 
+absl::optional<
+    envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelector::LbSubsetSelectorFallbackPolicy>
+SubsetLoadBalancer::tryFindSelectorFallbackPolicy(LoadBalancerContext* context) {
+  const Router::MetadataMatchCriteria* match_criteria = context->metadataMatchCriteria();
+  if (!match_criteria) {
+    return absl::nullopt;
+  }
+  const auto match_criteria_vec = match_criteria->metadataMatchCriteria();
+  SubsetSelectorMapPtr selectors = selectors_;
+  if (selectors == nullptr) {
+    return absl::nullopt;
+  }
+  for (uint32_t i = 0; i < match_criteria_vec.size(); i++) {
+    const Router::MetadataMatchCriterion& match_criterion = *match_criteria_vec[i];
+    const auto& subset_it = selectors->subset_keys.find(match_criterion.name());
+    if (subset_it == selectors->subset_keys.end()) {
+      // No subsets with this key (at this level in the hierarchy).
+      break;
+    }
+
+    if (i + 1 == match_criteria_vec.size()) {
+      // We've reached the end of the criteria, and they all matched.
+      return subset_it->second->fallback_policy;
+    }
+    selectors = subset_it->second;
+  }
+
+  return absl::nullopt;
+}
+
+HostConstSharedPtr SubsetLoadBalancer::chooseHostForSelectorFallbackPolicy(
+    const absl::optional<
+        envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelector::LbSubsetSelectorFallbackPolicy>&
+        fallback_policy,
+    LoadBalancerContext* context) {
+  if (!fallback_policy) {
+    return nullptr;
+  }
+  const auto fallback = fallback_policy.value();
+  if (fallback == envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelector::ANY_ENDPOINT &&
+      selector_fallback_subset_any_ != nullptr) {
+    HostConstSharedPtr host =
+        selector_fallback_subset_any_->priority_subset_->lb_->chooseHost(context);
+    if (host != nullptr) {
+      return host;
+    }
```

Comment:
nit: elide local variable and return return the result of chooseHost (same below).